### ISSUE TITLE
DataLoader worker Error

### ISSUE THREAD

**Issue by matany58** [chemprop/chemprop#150]
I downloaded the latest version of Chemprop today 

I used the prediction option  with large dataset ( about 5 million records) and  the error below showed up;

 raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 65618) exited unexpectedly

The prediction option is given below;

python predict.py --test_path <file path>  --checkpoint_path <model path> --num_workers 24  --preds_path <prediction save path)

**Comment by hesther** [comment]
Without the full error message, it is hard to tell what happened here. A few things to try:
1) Does the same error occur for a smaller dataset, too? You could prepare a dummy dataset with just a few lines to test.
2) When does the error occur - straight away or in the middle of predicting values for the dataset?
3) If no error occurs for a smaller dataset, it seems to be a memory issue and you can try splitting up the predictions in a few chunks (or working on a machine with more memory).
4) Could you provide the full error message? 
5) What operating system (Linux, MacOS, Windows) and Python Version are you on?

Thanks!

**Comment by matany58** [comment]
Hello Hesther;

The following is reply to your message;
1. It does not occur for smaller dataset
2. It occurs in the middle of predicting values for the dataset
3. The machine is Redhat linux OS version 7
4. The error is bellow;
Traceback (most recent call last):
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 986, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/multiprocessing/queues.py", line 104, in get
    if not self._poll(timeout):
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/multiprocessing/connection.py", line 257, in poll
    return self._poll(timeout)
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/multiprocessing/connection.py", line 414, in _poll
    r = wait([self], timeout)
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/multiprocessing/connection.py", line 921, in wait
    ready = selector.select(timeout)
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 53307) is killed by signal: Killed.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "predict.py", line 6, in <module>
    chemprop_predict()
  File "/research/sharedresources/leadDiscoveryInformaticsCenter/projects/ML/conda_envs/chemprop/chemprop/train/make_predictions.py", line 172, in chemprop_predict
    make_predictions(args=PredictArgs().parse_args())
  File "/research/sharedresources/leadDiscoveryInformaticsCenter/projects/ML/conda_envs/chemprop/chemprop/utils.py", line 437, in wrap
    result = func(*args, **kwargs)
  File "/research/sharedresources/leadDiscoveryInformaticsCenter/projects/ML/conda_envs/chemprop/chemprop/train/make_predictions.py", line 104, in make_predictions
    scaler=scaler
  File "/research/sharedresources/leadDiscoveryInformaticsCenter/projects/ML/conda_envs/chemprop/chemprop/train/predict.py", line 27, in predict
    for batch in tqdm(data_loader, disable=disable_progress_bar, leave=False):
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 517, in __next__
    data = self._next_data()
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1182, in _next_data
    idx, data = self._get_data()
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1148, in _get_data
    success, data = self._try_get_data()
  File "/home/manyanwu/.conda/envs/chemprop1/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 999, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 53307) exited unexpectedly

**Comment by hesther** [comment]
This seems to be a memory issue, so I assume you are running out of memory before all 5 million records are processed. I think the easiest way to fix this would be to split your input data in a few files, predict separately on them, and then concatenate the files (if you need all predictions in a single file). If (don't feel obliged to) you want to confirm this is a memory issue, you could run Chemprop in one tab (I assume you are using the command line) and monitoring the memory requirement in another e.g using "top", and you should see how the increasing memory demand causes the job to fail.

**Comment by hesther** [comment]
I will run some tests and report to you later on!

**Comment by hesther** [comment]
Just ran some tests on my (old, memory-limited) laptop, and could easily run a prediction on 5M compounds, tried with 1 or 24 workers (although it supports only 8 parallel processes, so the workers wait for resources all the time) with a single regression task, standard parameters (about 0.3M parameters). Is there anything memory-intensive about your model (e.g. multiclass classification with many classes, large ensemble, etc.) which I can test additionally? 

Furthermore, I noticed there is absolutely no speed improvement if employing multiple workers - this might be a bug (I guess the keyword num_workers is only used in training mode - have to check).

**Comment by matany58** [comment]
Hesther
Thanks for your help. I am actually predicting about I billion compounds, so I split it up into 20M compounds. I have no issue predicting with 5M or 10M compounds   but the issue arises when the compound is 20M. With the files split into 20M compounds, sometimes it was a success and at other time the data loader error occurs. Also there is nothing memory intensive about the prediction. I think two issues; how to increase the chemprop memory so it does not run out memory as my system has enough memory . Also how to increase the parallel loading of data

**Comment by hesther** [comment]
I agree - currently parts of data loading and preprocessing (e.g. validating) are sequential. Especially validation takes up a significant amount of time and could easily be run in parallel. To illustrate, I ran a few tests (default model of esol dataset, no ensembling, predicting on multiples of esol dataset, 2,2 GHz Quad-Core no GPU used):
```
                               num_workers=4             num_workers=0
                              -----------------         ---------------
                              0.1M    1M     3M         0.1M    1M         

get_data (read from file)     0:01    0:10   0:31       0:01    0:10
validate smiles               0:15    2:24   7:22       0:14    2:28
prediction                    0:36    6:17  19:30       1:04   10:28
total                         0:52    8:59  27:42       1:21   13:12
```
I think we should decrease the time for validating (currently about 25% of total time) by parallelising (and it would be easy to do so).  I can look into it in the near future (but might be a few weeks, since I have some other tasks to work on currently) - if someone else is free before that, please feel free to work on it. I also want to reiterate what was already said in a previous issue: There is no benefit of running one large prediction over multiple small ones. Actually, if you look at the timings from my dummy benchmark above, it is faster to predict in chunks. However, the timings depend a lot on the model and data details (e.g. larger molecules take longer to featurize), so yours might look very different. So, I would also recommend to break up your data into smaller chunks (1M or 5M or 10M instead of 20M).

Also, I was a bit confused: Did you predict on 5M or 20M (your recent comment said 20M, the initial question 5M). 

Last but not least, just out of curiosity, since you said in a previous issue that 20M compounds took days to predict on a GPU: Could you solve this problem? Depending in your model and ensemble size, you might end up with quite different timings than I did above, but it certainly should not take days (in my case, extrapolating would suggest about 180mins, and that was on a slow laptop, no GPU, but again, it depends on your data, model and ensemble size).

**Comment by hesther** [comment]
Closing this due to no recent activity. @matany58, please feel free to reopen or post another issue if you need further assistance!
